src/core/programs.rkt

```diff
@@ -116,39 +129,67 @@
     [(? literal?) '()]
     [(? number?) '()]
     [(? variable?) (list prog)]
-    [`(,op ,args ...) (remove-duplicates (append-map free-variables args))]))
+    [(approx spec impl) (remove-duplicates (append (free-variables spec) (free-variables impl)))]
+    [(list _ args ...) (remove-duplicates (append-map free-variables args))]))
 
 (define (replace-vars dict expr)
-  (cond
-    [(dict-has-key? dict expr) (dict-ref dict expr)]
-    [(list? expr) (cons (replace-vars dict (car expr)) (map (curry replace-vars dict) (cdr expr)))]
-    [#t expr]))
+  (let loop ([expr expr])
+    (match expr
+      [(? literal?) expr]
+      [(? number?) expr]
+      [(? symbol?) (dict-ref dict expr expr)]
+      [(approx impl spec) (approx (loop impl) (loop spec))]
+      [(list op args ...) (cons op (map loop args))])))
+
+;; For any LImpl expression, removes any approx nodes.
+;; WARN: this is an irreversible spec-altering transformation.
+(define (remove-approx expr)
+  (match expr
+    [(? literal?) expr]
+    [(? number?) expr]
+    [(? symbol?) expr]
+    [(approx _ impl) (remove-approx impl)]
+    [(list op args ...) `(,op ,@(map remove-approx args))]))
 
 (define location? (listof natural-number/c))
 
 (define/contract (location-do loc prog f)
   (-> location? expr? (-> expr? expr?) expr?)
-  (cond
-    [(null? loc) (f prog)]
-    [(not (pair? prog)) (error "Bad location: cannot enter " prog "any further.")]
-    [#t
-     ; Inlined loop for speed
-     (let loop ([idx (car loc)] [lst prog])
-       (if (= idx 0)
-           (cons (location-do (cdr loc) (car lst) f) (cdr lst))
-           (cons (car lst) (loop (- idx 1) (cdr lst)))))]))
+  (define (invalid! where loc)
+    (error 'location-do "invalid location `~a` for `~a` in `~a`" loc where prog))
+
+  (let loop ([prog prog] [loc loc])
+    (match* (prog loc)
+      [(_ (? null?)) (f prog)]
+      [((or (? literal?) (? number?) (? symbol?)) _) (invalid! prog loc)]
+      [((approx spec impl) (cons idx rest)) ; approx nodes
+       (case idx
+         [(1) (approx (loop spec rest) impl)]
+         [(2) (approx spec (loop impl rest))]
+         [else (invalid! prog loc)])]
+      [((list op args ...) (cons idx rest)) ; operator
+       (let seek ([elts (cons op args)] [idx idx])
+         (cond
+           [(= idx 0) (cons (loop (car elts) rest) (cdr elts))]
+           [(null? elts) (invalid! prog loc)]
+           [else (cons (car elts) (seek (cdr elts) (sub1 idx)))]))]
+      [(_ _) (invalid! prog loc)])))
 
 (define/contract (location-get loc prog)
   (-> location? expr? expr?)
   ; Clever continuation usage to early-return
   (let/ec return (location-do loc prog return)))
 
-(define/contract (replace-expression haystack needle needle*)
+(define/contract (replace-expression expr to from)
```

Comment:
Bizarre argument names… why not from to?

Reply:
Hmm weird